[IMP] make customize_show views part of multi-theme

yelizariev opened the pull request:
The PR allows making customization independently per website
![multi-customization](https://user-images.githubusercontent.com/186131/38645757-a921d322-3dfe-11e8-89e4-63caeb4329aa.png)

yelizariev commented on website_multi_theme/data/themes_default.xml:
```diff
@@ -0,0 +1,25 @@
+     License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html). -->
+<odoo>
+    <!-- Default theme. It's basically a workaround for "Multi-customization"
+         feature which requires theme be set -->
```
I don't see other solution. At least until we make refactoring discussed #409

yelizariev commented on website_multi_theme/models/ir_ui_view.py:
```diff
@@ -22,3 +24,29 @@ class IrUiView(models.Model):
+        string="Copied from",
+        help="View from where this one was copied for multi-website"
+    )
+    copy_ids = fields.One2many(
```
Those two fields are left after initial draft of this update, but I think it could useful later

yelizariev commented on website_multi_theme/models/website.py:
```diff
@@ -82,18 +85,15 @@ def _duplicate_view_for_website(self, pattern, xmlid, override_key):
-                xmlid,
-                self.display_name,
-            ),
+            "name": '%s (Multi-Website)' % pattern.name,
```
It's renamed, because this name is used in Customize menu (grayed text)

yelizariev replied:
Maybe, the name can be changed to ``'%s (Website #%s)' % (pattern.name,  self.id)`` , because **Multi-website** sounds like **the same per different websites**

yelizariev replied:
I don't want to use ``self.display_name``, because website could be renamed

yelizariev commented on website_multi_theme/models/website.py:
```diff
@@ -82,18 +85,15 @@ def _duplicate_view_for_website(self, pattern, xmlid, override_key):
-            ),
+            "name": '%s (Multi-Website)' % pattern.name,
             "website_id": self.id,
+            "origin_view_id": pattern.id,
```
It's kind of replacement to removed data in name

yelizariev commented on website_multi_theme/models/website.py:
```diff
@@ -178,12 +178,12 @@ def _multi_theme_activate(self):
+                    # used
+                    copied_parent = self._find_duplicate_view_for_website(
+                        parent_view, website
+                    )
```
It's not directly part of this update, though there is a theoretical situation when several Customized views inherit each other, so they need to use a copied view too. But I need to make real multi-footer in my module ``website_multi_company`` to apply some extension to make ``website.footer_custom`` copyable too

yelizariev commented on website_multi_theme/models/website_theme.py:
```diff
@@ -36,20 +36,48 @@ class WebsiteTheme(models.Model):
+
+            if refs or one.asset_ids:
+                # add common_refs only for installed themes
+                refs += common_refs
```
This is the key idea of the update

yajo replied:
In general I recommend to use `|=`, to avoid duplicates in the recordsets.

yelizariev commented on website_multi_theme/templates/patterns.xml:
```diff
@@ -6,7 +6,7 @@
         id="layout_pattern"
         inherit_id="website.layout"
-        name="Template for replacing assets_frontend"
+        name="Main layout"
```
it's renamed because it's used in Customize menu

yelizariev commented on website_multi_theme/models/website.py:
```diff
@@ -82,18 +85,15 @@ def _duplicate_view_for_website(self, pattern, xmlid, override_key):
         result = pattern.copy({
             "active": pattern.was_active,
             "arch_fs": False,
-            "customize_show": False,
```
@Yajo is there any reason why this was set to False? Do you remember the case when it's required?

yelizariev commented on website_multi_theme/README.rst:
```diff
@@ -117,7 +117,7 @@ Known issues / Roadmap
 * If you install any of the supported themes after installing this addon, you
   will have to press *Reload* in the website config wizard to make it notice
-  the change.
+  the change. The same is true when you install any module with ``customize_show="True"``
```
Tests are not passed in travis due to this new requirement: *Reload* must be done after installing  website_blog_category https://github.com/OCA/website/blob/10.0/website_blog_category/templates/website_blog_template.xml#L25

What can we do here? @Yajo @pedrobaeza

yajo replied:
What about calling reload from the last demo data file read?

yelizariev replied:
Do you mean demo file of **website_blog_category**?

yelizariev replied:
I'd prefer to Reload after creating new ``ir.ui.view`` record with ``customize_show`` (via extending ``create`` method). What do you think, @pedrobaeza ?

yajo replied:
That would make sense, yes.

I was talking about demo data file included in this addon. Or even maybe a normal data file.

yelizariev commented on website_multi_theme/models/website.py:
```diff
@@ -85,15 +88,10 @@ def _duplicate_view_for_website(self, pattern, xmlid, override_key):
         result = pattern.copy({
             "active": pattern.was_active,
             "arch_fs": False,
-            "customize_show": False,
```
*Copied message to don't miss it:*

@Yajo is there any reason why this was set to False? Do you remember the case when it's required?

yajo replied:
Can't remember, honestly :roll_eyes:

yelizariev commented:
When you review PR yourself

![dog-walking-itself-1](https://user-images.githubusercontent.com/186131/38648354-35be0368-3e0a-11e8-8b45-fc67859412b3.gif)

yajo commented on website_multi_theme/models/website.py:
```diff
@@ -171,19 +169,15 @@ def _multi_theme_activate(self):
-
                     parent_view = copied_view.inherit_id
-                    parent_view_module = parent_view.model_data_id.module
                     copied_parent = None
```
This line makes no sense now, isn't it?

yajo commented on website_multi_theme/models/ir_ui_view.py:
```diff
@@ -13,12 +14,36 @@
 class IrUiView(models.Model):
     _inherit = 'ir.ui.view'
 
-    multi_theme_generated = fields.Boolean(
```
You need a migration script to fill `origin_view_id` from preexisting views that had this boolean set to true.

yelizariev replied:
I'll do that

yajo commented on website_multi_theme/models/website.py:
```diff
@@ -25,11 +25,14 @@ class Website(models.Model):
         inverse_name="website_id",
-        domain=[("multi_theme_generated", "=", True),
+        domain=[("origin_view_id", "!=", False),
                 "|", ("active", "=", True), ("active", "=", False)],
```
Maybe we should replace this by `context={"active_test": True}`

yelizariev replied:
I cannot find an example of using such trick. Are you sure that context is used when we read the field in python via ``.multi_theme_view_ids`` and other cases? I'd prefer to keep as it's now

yajo replied:
OK no problem, it was just an idea

yajo commented on website_multi_theme/models/website_theme.py:
```diff
@@ -36,20 +36,48 @@ class WebsiteTheme(models.Model):
-            ])
-            existing = frozenset(one.mapped("asset_ids.name"))
+
+        common_refs = self.env["ir.model.data"]
```
This line is not needed; the next sentence will return an empty recordset when nothing is found.

yelizariev replied:
The idea was to allow in possible next updates don't change git blame much.

For example,

Version1

common_refs = self.env["ir.model.data"]
common_refs |= A

Version 2

common_refs = self.env["ir.model.data"]
common_refs |= A
common_refs |= B

Version 3 

common_refs = self.env["ir.model.data"]
common_refs |= A1
common_refs |= B
common_refs |= C

Version 5


common_refs = self.env["ir.model.data"]
common_refs |= B
common_refs |= C

yajo commented on website_multi_theme/models/website_theme.py:
```diff
@@ -87,6 +115,12 @@ class WebsiteThemeAsset(models.Model):
              "website, and disabled otherwise. Usually used to load assets.",
     )
 
+    auto = fields.Boolean(
```
Move fields above, just like [explained here](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#models).

yelizariev replied:
Do you mean above ``_sql_constraints``? It's actually not explained in the guidelines. It mentions only constrain methods. While ``_sql_constraints`` seems to be part of 

    Private attributes (_name, _description, _inherit, ...)

yajo replied:
Yikes! forget this. I think I didn't expand the diff in between :laughing:

yajo commented on website_multi_theme/views/ir_ui_view.xml:
```diff
@@ -0,0 +1,26 @@
+                    <separator string="Copies"/>
+                    <field name="copy_ids"/>
+                </page>
+                <page string="Multi-website" attrs="{'invisible': [('origin_view_id', '=', False)]}">
```
2 pages called the same?

Better 1 page and 2 groups.

Remember to add a `name="multiwebsite"`, to enable possible inheritance.

yelizariev replied:
The idea is don't pass attributes twice in each node here:

                    <separator string="Copies"/>
                    <field name="copy_ids"/>

yelizariev replied:
I can add different names, if it's the concern

    name="multiwebsite_origin_view"
    name="multiwebsite_copied_view"

yajo commented on website_multi_theme/wizards/website_config_settings_view.xml:
```diff
@@ -15,7 +15,7 @@
                 <label for="multi_theme_id"/>
                 <div>
-                    <field name="multi_theme_id" class="oe_inline"/>
+                    <field name="multi_theme_id" class="oe_inline" required="1"/>
```
Why not require it in the model?

yelizariev replied:
Because we have default website, which doesn't have multi_theme_id on installation and we cannot set it on installation, because *Default theme* is created after loading python files

yajo replied:
It makes sense. Could you add a note here then please? It's not obvious...

yajo commented on website_multi_theme/README.rst:
```diff
@@ -137,6 +137,7 @@ Contributors
 * Rafael Blasco <rafael.blasco@tecnativa.com>
 * Antonio Espinosa <antonio.espinosa@tecnativa.com>
 * Jairo Llopis <jairo.llopis@tecnativa.com>
+* Ivan Yelizariev <https://it-projects.info/team/yelizariev>
```
I can't believe you weren't here before!

yelizariev commented:
Rebased and added some fixes. Few comments from @Yajo's review are subject to discuss. I'll squash commits once we close all concerns or one can do it on merging (just fist commit message has to be keeped: ``[IMP] make customize_show views part of multi-theme``)

yelizariev commented:
About this idea:

> All views with customize_show=True are considered multiwebsite views. I'm not sure if we should instead duplicate all views belonging to addons that depend on website; after all, a user usually can customize any view in a website... :thinking:

The possible problem here is that those changes will be destroyed after switching a website to new theme. So, it's better to share them, than letting user face data lost

yelizariev commented:
I'll do this later

- [x] a migration script to fill origin_view_id from preexisting views that had this boolean set to true.

rafaelbn commented:
Thank you @yelizariev for this great contribution! :smile: :clap: :clap: :clap:

yelizariev commented:
I added migration scripts and squashed the PR

Previous history:
https://github.com/yelizariev/website/commits/10.0-website_multi_company-multi-customize-before-squashing

yelizariev commented:
And another commit  https://github.com/yelizariev/website/commit/1984fbf5566e7c21dc1e7942cfaf704badc57d4c
(squashed)

yelizariev commented:
One more update from @Yajo's review  https://github.com/yelizariev/website/commit/0fd0ad3809aaab98eb58085f4836d8f268387964

and there is still [the question](https://github.com/OCA/website/pull/443#discussion_r180923312) about travis tests. Can we do something about it or just left it as is

ilmir-k commented:
@yelizariev 
During the tests I found out wrong behavior. When we use ``Default theme`` or ``Multiwebsite Demo Theme`` for websites everything is ok, but if we install and set **Bootswatch Theme** we miss the options for ``Customize``

![missed_options](https://user-images.githubusercontent.com/14342645/38888692-e56b7e1e-4295-11e8-9d98-54f985dace3c.png)

yelizariev commented:
I think the error was introduced in https://github.com/OCA/website/commit/a4ed42fe08278cff4ec618e785c64dedd35f9d5c#diff-93d51900ef6c23af9b0a13e1278dfad5R45 by @Yajo 

How to reproduce without this updates:

* Install *Bootswatch Theme*
* Click reload twice
* RESULT: *Multi-theme Bootswatch Theme* is disappeared from the list

yelizariev commented:
@Yajo could you check and fix?

yajo commented:
See #447

yelizariev commented:
Rebased
```

ivann@ivann-360:/mnt/files/odoo/odoo-10.0/website$ git fetch OCA  pull/443/head
De https://github.com/OCA/website
 * branch            refs/pull/443/head -> FETCH_HEAD
ivann@ivann-360:/mnt/files/odoo/odoo-10.0/website$ git diff FETCH_HEAD..HEAD
diff --git a/website_multi_theme/models/website_theme.py b/website_multi_theme/models/website_theme.py
index 45f3ea4..a0190f2 100644
--- a/website_multi_theme/models/website_theme.py
+++ b/website_multi_theme/models/website_theme.py
@@ -65,10 +65,11 @@ class WebsiteTheme(models.Model):
                 # add common_refs only for installed themes
                 refs |= common_refs
 
-            views = self.env["ir.ui.view"].search([
-                ("id", "in", refs.mapped("res_id")),
-                ("type", "=", "qweb"),
-            ])
+            views = self.env["ir.ui.view"].with_context(active_test=False) \
+                .search([
+                    ("id", "in", refs.mapped("res_id")),
+                    ("type", "=", "qweb"),
+                ])
             existing = frozenset(
                 one
                 .mapped("asset_ids")

```

yelizariev commented:
Update for the last question: https://github.com/yelizariev/website/commit/3b61c5d086ba50a9bd914c4750ec4410f0e4699d

It took more time than I [thought](https://github.com/OCA/website/pull/443#discussion_r182094341)

@Yajo your review is very needed

yelizariev commented:
Other updates

LINT: https://github.com/yelizariev/website/commit/52b962fa313c9d207d2116b629ec1f97332b37ce
CI fix:  https://github.com/yelizariev/website/commit/6c5a852892339988162b6db4d6182b5c71be936a

rafaelbn commented:
Please @Yajo could you review here? Thank you!

yajo commented on website_multi_theme/models/ir_ui_view.py:
```diff
@@ -13,12 +15,36 @@
         help="Indicates if the view was originally active before converting "
              "the single website theme that owns it to multi website mode.",
     )
+
```
I'd remove this white line

yajo commented on website_multi_theme/models/website.py:
```diff
@@ -125,15 +126,21 @@ def _multi_theme_activate(self):
                 )
-                website.multi_theme_view_ids.unlink()
-                continue
+                if not default_theme:
```
If I'm not wrong, this `if` will never pass because the above call to `ref()` doesn't have a `raise_if_not_found=False`.

OTOH I'd use the try-except-else option instead.

yajo commented on website_multi_theme/models/website_theme.py:
```diff
@@ -36,25 +37,54 @@ class WebsiteTheme(models.Model):
-                ("model", "=", "ir.ui.view"),
-            ])
+
+        common_refs = self.env["ir.model.data"]
```
You can delete this, the search below will return an empty recordset on his own if no results are found.

yelizariev replied:
My explanation is here: https://github.com/OCA/website/pull/443#discussion_r181154724
You probably miss it. What do you think about it?

yelizariev commented:
Tiny updates https://github.com/yelizariev/website/commit/e028b325a1329bad03542f29b814c151b2ca4c48

yelizariev commented:
@pedrobaeza ?

pedrobaeza commented on website_multi_theme/__manifest__.py:
```diff
@@ -1,12 +1,13 @@
+    "version": "10.0.1.2.0",
     "category": "Website",
     "website": "https://www.tecnativa.com",
     "author": "Tecnativa, Odoo Community Association (OCA)",
```
@yelizariev your company deserves to be here

pedrobaeza commented on website_multi_theme/models/ir_model_data.py:
```diff
@@ -0,0 +1,37 @@
+    def create(self, vals):
+        res = super(IrModelData, self).create(vals)
+        if vals.get('model') == 'ir.ui.view' \
+           and not self.env.context.get('duplicate_view_for_website'):
```
Incorrect indendation and line split

yelizariev replied:
Multi-line if-block is my weak side :disappointed:

pedrobaeza replied:
The best is to enclose in parenthesis and always indent second line 2 levels more:
```python
if (vals.get('model') == 'ir.ui.view' and
        not self.env.context.get('duplicate_view_for_website')):
```

yelizariev replied:
I'd prefer to put logical operator at the beginning of the line

yelizariev replied:
Oh, it's W503 requirement. OK

yelizariev replied:
I'm not the only one who thinks so.

W503 is ignored by default in latest version of pycodestyle.

https://github.com/OCA/maintainer-quality-tools/issues/545

yelizariev replied:
I said wrong. W504 is ignored by default

yelizariev replied:
But people say that W504 is more logical

pedrobaeza replied:
Well, we can discuss about that in the issue, but right now it's W503 what is used.

pedrobaeza commented on website_multi_theme/models/ir_ui_view.py:
```diff
@@ -13,12 +15,35 @@
+        "origin_view_id",
+        string="Copies",
+        readonly=True,
+        help="Duplicates of this view"
```
I would prefer a more explicit name: `multitheme_copy_ids` and a longer explanation in help.

pedrobaeza commented on website_multi_theme/models/ir_model_data.py:
```diff
@@ -0,0 +1,37 @@
+           and not self.env.context.get('duplicate_view_for_website'):
+            view = self.env['ir.ui.view'].browse(vals.get('res_id'))
+            if view.customize_show:
+                # Reload theme to make new customize_show available right after
```
Put this explanation in the method docstring instead of here, as it's the purpose of the override.

pedrobaeza commented on website_multi_theme/models/ir_ui_view.py:
```diff
@@ -13,12 +15,35 @@
+        views = super(IrUiView, self)._customize_template_get_views(
+            key, full=full, bundles=bundles
+        )
+
```
These empty lines don't add anything useful and PEP8 discourages its use. Better to strip them.

pedrobaeza commented on website_multi_theme/models/website.py:
```diff
@@ -25,11 +26,14 @@ class Website(models.Model):
         comodel_name='website.theme',
         domain=[("asset_ids.view_id", "!=", False)],
         help="Multiwebsite-compatible theme for this website",
+        require=True,
```
Please don't add require fields in inherited classes, as this will make that tests that create records from this model fail when this dependent module is not loaded. Instead, put the require on the view.

yelizariev replied:
I believe it's true for existing fields only, while here we create a new field.

pedrobaeza replied:
No, it's the same. The sequence is the following in MQT:

- This module is installed along with all the testable ones and its dependencies.
- Then the SQL NOT NULL constrained is added.
- On the second pass, modules are being tested while being loaded.
- Any module adding a website will have the not null error until this module is loaded.

yelizariev replied:
I got it. You are right. 
It seems to be an odoo framework issue. Is there a way to fix it in odoo core?

pedrobaeza commented on website_multi_theme/models/website_theme.py:
```diff
@@ -36,25 +37,54 @@ class WebsiteTheme(models.Model):
+                               ('website_id', '=', False),
+                               ('customize_show', '=', True),
+                           ]).mapped('model_data_id')
+
```
Don't add too much empty lines...

yelizariev commented:
New updates: https://github.com/yelizariev/website/commit/3a24f70032d6db872b795257eb91b698d9dde16d

yelizariev commented:
And update about ``required=True`` https://github.com/yelizariev/website/commit/b70f910cbcda7a8f02fd9fcbf10db013317bd470

yelizariev commented:
I suppose this is the last question here:

    website_multi_theme/models/ir_model_data.py:33:17: W503 line break before binary operator

I made an issue in MQT https://github.com/OCA/maintainer-quality-tools/issues/545

pedrobaeza commented:
But that's because you haven't put the `and` in the previous line.

yelizariev commented:
I know, but I'd prefer to put ``and`` at the beginning of the line. 
![why](https://user-images.githubusercontent.com/186131/39143264-cf6d6aa4-4746-11e8-93a7-3ba717488bb4.png)

pedrobaeza commented:
Yeah, but if you want this merged before discussing about the change, you have to fix it. If not, we have to wait until the discussion is solved.

yelizariev commented:
I'd prefer to merge it now.
(Travis is green now)

rafaelbn commented:
Thanks @yelizariev ! :clap: :clap:
